packages/ra-core/src/controller/details/useCreateContext.tsx

```diff
@@ -14,21 +33,54 @@ export const useCreateContext = <
         CreateContext
     );
 
-    if (!context.resource) {
-        /**
-         * The element isn't inside a <CreateContext.Provider>
-         * To avoid breakage in that case, fallback to props
-         *
-         * @deprecated - to be removed in 4.0
-         */
-        if (process.env.NODE_ENV !== 'production') {
-            console.log(
-                "Create components must be used inside a <CreateContext.Provider>. Relying on props rather than context to get Create data and callbacks is deprecated and won't be supported in the next major version of react-admin."
-            );
-        }
-
-        return props;
-    }
-
-    return context;
+    // Props take precedence over the context
+    // @ts-ignore
+    return props != null
+        ? merge({}, context, extractCreateContextProps(props))
```

Comment:
You must memoize this, otherwise the hook will return a new object at each render.